src/tokenizer/shingle_filter.rs

```diff
@@ -0,0 +1,207 @@
+use super::{BoxTokenStream, Token, TokenFilter, TokenStream};
+
+#[derive(Clone)]
+pub struct FixedShingleFilter {
+    size: usize,
+    separator: char,
+}
+
+impl FixedShingleFilter {
+    pub fn new(size: usize) -> Self {
+        FixedShingleFilter::new_with_separator(size, '_')
+    }
+
+    pub fn new_with_separator(size: usize, separator: char) -> Self {
+        FixedShingleFilter { size, separator }
+    }
+}
+
+impl TokenFilter for FixedShingleFilter {
+    fn transform<'a>(&self, token_stream: BoxTokenStream<'a>) -> BoxTokenStream<'a> {
+        BoxTokenStream::from(ShingleFilterStream {
+            size: self.size,
+            separator: self.separator,
+            buffer: String::with_capacity(100),
+            pendings: Vec::with_capacity(self.size - 1),
+            tail: token_stream,
+            original_token_position: 0,
+        })
+    }
+}
+
+pub struct ShingleFilterStream<'a> {
+    size: usize,
+    separator: char,
+    buffer: String,
+    pendings: Vec<Token>,
+    tail: BoxTokenStream<'a>,
+    original_token_position: usize,
+}
+
+impl<'a> ShingleFilterStream<'a> {
+    fn is_window_filled(&self) -> bool {
+        self.pendings.len() == self.size - 1
+    }
+
+    fn push_token(&mut self, mut token: Token) {
+        if self.is_window_filled() {
+            self.pendings.remove(0);
+        }
+        token.position = self.original_token_position;
+        self.pendings.push(token);
+    }
+}
+
+fn make_shingle(token: &mut Token, pendings: &Vec<Token>, buffer: &mut String, separator: char) {
+    buffer.clear();
+    buffer.push_str(token.text.as_str());
```

Comment:
nitpick
That just removes one line and one token copy so it does not matter much...
But mem::swap could work too here.

Reply:
(please do not address this comment. the code is fine as is.)